File: Src/SharedMemory/examples/cross_language_test.cs
/*
 * 跨语言共享内存测试 - C#示例
 * 测试与C++和Python的跨语言通信
 */

using System;
using System.Collections.Generic;
using System.Threading;
using BitRPC.SharedMemory;

// 测试数据结构
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct TestData
{
    public int Id;
    public double Value;
    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 64)]
    public string Message;
    public ulong Timestamp;
}

public class CrossLanguageTestProducer
{
    private readonly string name;
    private volatile bool running;
    private SharedMemoryProducer producer;
    private Thread sendThread;

    public CrossLanguageTestProducer(string name = "CrossLangTest")
    {
        this.name = name;
        this.running = false;
    }

    ~CrossLanguageTestProducer()
    {
        Stop();
    }

    public bool Start()
    {
        if (running) return true;

        try
        {
            // 创建生产者
            producer = SharedMemoryFactory.CreateProducer(name, 1024 * 1024); // 1MB buffer
            if (producer == null)
            {
                Console.WriteLine("✗ Failed to create producer");
                return false;
            }

            Console.WriteLine("✓ C# Producer connected to shared memory: " + name);
            running = true;

            // 启动发送线程
            sendThread = new Thread(SendLoop);
            sendThread.IsBackground = true;
            sendThread.Start();

            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error starting producer: {ex.Message}");
            return false;
        }
    }

    public void Stop()
    {
        if (!running) return;

        running = false;

        if (sendThread != null && sendThread.IsAlive)
        {
            sendThread.Join();
        }

        producer?.Dispose();
        Console.WriteLine("✓ C# Producer stopped");
    }

    public bool IsRunning => running;

    private void SendLoop()
    {
        int counter = 0;

        while (running)
        {
            try
            {
                // 发送文本消息
                string textMessage = $"Hello from C#! Message #{counter}";
                if (producer.SendString(textMessage))
                {
                    Console.WriteLine($"Sent text: {textMessage}");
                }

                Thread.Sleep(100);

                // 发送结构化数据
                var data = new TestData
                {
                    Id = counter,
                    Value = 2.71828 * counter,
                    Message = $"C# Data #{counter}",
                    Timestamp = (ulong)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
                };

                var message = new SharedMemoryMessage(MessageType.Data, StructureToBytes(data));
                message.SetFlag(MessageFlags.Urgent);

                if (producer.SendMessage(message))
                {
                    Console.WriteLine($"Sent structured data: id={data.Id}, value={data.Value}");
                }

                // 发送心跳
                if (counter % 10 == 0)
                {
                    if (producer.SendMessage(MessageType.Heartbeat, Array.Empty<byte>()))
                    {
                        Console.WriteLine("Sent heartbeat");
                    }
                }

                counter++;

                // 等待一段时间
                Thread.Sleep(500);

            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in send loop: {ex.Message}");
                Thread.Sleep(1000);
            }
        }
    }

    private static byte[] StructureToBytes<T>(T structure) where T : struct
    {
        int size = Marshal.SizeOf<T>();
        byte[] arr = new byte[size];
        IntPtr ptr = Marshal.AllocHGlobal(size);
        try
        {
            Marshal.StructureToPtr(structure, ptr, true);
            Marshal.Copy(ptr, arr, 0, size);
            return arr;
        }
        finally
        {
            Marshal.FreeHGlobal(ptr);
        }
    }
}

public class CrossLanguageTestConsumer
{
    private readonly string name;
    private volatile bool running;
    private SharedMemoryConsumer consumer;
    private Thread receiveThread;

    public CrossLanguageTestConsumer(string name = "CrossLangTest")
    {
        this.name = name;
        this.running = false;
    }

    ~CrossLanguageTestConsumer()
    {
        Stop();
    }

    public bool Start()
    {
        if (running) return true;

        try
        {
            // 创建消费者
            consumer = SharedMemoryFactory.CreateConsumer(name, 1024 * 1024); // 1MB buffer
            if (consumer == null)
            {
                Console.WriteLine("✗ Failed to create consumer");
                return false;
            }

            Console.WriteLine("✓ C# Consumer connected to shared memory: " + name);
            running = true;

            // 启动接收线程
            receiveThread = new Thread(ReceiveLoop);
            receiveThread.IsBackground = true;
            receiveThread.Start();

            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error starting consumer: {ex.Message}");
            return false;
        }
    }

    public void Stop()
    {
        if (!running) return;

        running = false;

        if (receiveThread != null && receiveThread.IsAlive)
        {
            receiveThread.Join();
        }

        consumer?.Dispose();
        Console.WriteLine("✓ C# Consumer stopped");
    }

    public bool IsRunning => running;

    private void ReceiveLoop()
    {
        while (running)
        {
            try
            {
                // 尝试接收消息
                if (consumer.ReceiveMessage(out var message, 100))
                {
                    Console.WriteLine($"Received message: type={message.MessageType}, size={message.Payload.Length} bytes");

                    if (message.MessageType == MessageType.Data && message.Payload.Length == Marshal.SizeOf<TestData>())
                    {
                        var data = BytesToStructure<TestData>(message.Payload);
                        Console.WriteLine($"  Parsed data: id={data.Id}, value={data.Value}, message={data.Message}, timestamp={data.Timestamp}");
                    }
                    else if (message.MessageType == MessageType.Heartbeat)
                    {
                        Console.WriteLine("  Received heartbeat from producer");
                    }
                    else if (message.MessageType == MessageType.Control)
                    {
                        string text = System.Text.Encoding.UTF8.GetString(message.Payload);
                        Console.WriteLine($"  Received control message: {text}");
                    }

                    continue;
                }

                // 尝试接收字符串
                if (consumer.ReceiveString(out var text, 100))
                {
                    Console.WriteLine($"Received string: {text}");
                    continue;
                }

                // 尝试接收原始数据
                if (consumer.Receive(out var data, 100))
                {
                    Console.WriteLine($"Received raw data: {data.Length} bytes");
                    continue;
                }

                // 如果没有数据，短暂休眠
                Thread.Sleep(50);

            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in receive loop: {ex.Message}");
                Thread.Sleep(1000);
            }
        }
    }

    private static T BytesToStructure<T>(byte[] bytes) where T : struct
    {
        IntPtr ptr = Marshal.AllocHGlobal(bytes.Length);
        try
        {
            Marshal.Copy(bytes, 0, ptr, bytes.Length);
            return Marshal.PtrToStructure<T>(ptr);
        }
        finally
        {
            Marshal.FreeHGlobal(ptr);
        }
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        Console.WriteLine("BitRPC Cross-Language Shared Memory Test (C#)");
        Console.WriteLine("============================================");

        string mode = "producer"; // 默认为生产者
        string name = "CrossLangTest";

        // 解析命令行参数
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "--mode" && i + 1 < args.Length)
            {
                mode = args[i + 1];
                i++;
            }
            else if (args[i] == "--name" && i + 1 < args.Length)
            {
                name = args[i + 1];
                i++;
            }
            else if (args[i] == "--help" || args[i] == "-h")
            {
                Console.WriteLine("Usage: Program.exe [options]");
                Console.WriteLine("Options:");
                Console.WriteLine("  --mode MODE    Set mode (producer/consumer)");
                Console.WriteLine("  --name NAME    Set shared memory name");
                Console.WriteLine("  --help, -h     Show this help");
                return;
            }
        }

        Console.WriteLine($"Mode: {mode}");
        Console.WriteLine($"Shared Memory Name: {name}");
        Console.WriteLine();

        try
        {
            if (mode == "producer")
            {
                using (var producer = new CrossLanguageTestProducer(name))
                {
                    if (!producer.Start())
                    {
                        Console.WriteLine("Failed to start producer");
                        return;
                    }

                    Console.WriteLine("Producer running. Press Ctrl+C to stop...");

                    // 等待用户中断
                    while (producer.IsRunning)
                    {
                        Thread.Sleep(1000);
                    }
                }
            }
            else if (mode == "consumer")
            {
                using (var consumer = new CrossLanguageTestConsumer(name))
                {
                    if (!consumer.Start())
                    {
                        Console.WriteLine("Failed to start consumer");
                        return;
                    }

                    Console.WriteLine("Consumer running. Press Ctrl+C to stop...");

                    // 等待用户中断
                    while (consumer.IsRunning)
                    {
                        Thread.Sleep(1000);
                    }
                }
            }
            else
            {
                Console.WriteLine($"Invalid mode: {mode}");
                Console.WriteLine("Supported modes: producer, consumer");
                return;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            Console.WriteLine($"Stack trace: {ex.StackTrace}");
            return;
        }

        Console.WriteLine("Test completed successfully!");
    }
}

File: Src/SharedMemory/examples/cross_language_test.cpp
/*
 * 跨语言共享内存测试 - C++生产者示例
 * 测试与C#和Python的跨语言通信
 */

#include "../shared_memory_api.h"
#include <iostream>
#include <thread>
#include <chrono>
#include <string>

using namespace bitrpc::shared_memory;

// 测试数据结构
#pragma pack(push, 1)
struct TestData {
    int id;
    double value;
    char message[64];
    uint64_t timestamp;
};
#pragma pack(pop)

class CrossLanguageTestProducer {
public:
    CrossLanguageTestProducer(const std::string& name = "CrossLangTest")
        : name_(name), running_(false) {
    }

    ~CrossLanguageTestProducer() {
        stop();
    }

    bool start() {
        if (running_) {
            return true;
        }

        // 创建生产者
        producer_ = create_producer(name_, 1024 * 1024); // 1MB buffer
        if (!producer_) {
            std::cerr << "Failed to create producer" << std::endl;
            return false;
        }

        std::cout << "✓ C++ Producer connected to shared memory: " << name_ << std::endl;
        running_ = true;

        // 启动发送线程
        send_thread_ = std::make_unique<std::thread>(&CrossLanguageTestProducer::send_loop, this);

        return true;
    }

    void stop() {
        if (!running_) {
            return;
        }

        running_ = false;

        if (send_thread_ && send_thread_->joinable()) {
            send_thread_->join();
        }

        producer_.reset();
        std::cout << "✓ C++ Producer stopped" << std::endl;
    }

    bool is_running() const {
        return running_;
    }

private:
    void send_loop() {
        int counter = 0;

        while (running_) {
            try {
                // 发送文本消息
                std::string text_message = "Hello from C++! Message #" + std::to_string(counter);
                if (producer_->send_string(text_message)) {
                    std::cout << "Sent text: " << text_message << std::endl;
                }

                std::this_thread::sleep_for(std::chrono::milliseconds(100));

                // 发送结构化数据
                TestData data;
                data.id = counter;
                data.value = 3.14159 * counter;
                snprintf(data.message, sizeof(data.message), "C++ Data #%d", counter);
                data.timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
                    std::chrono::system_clock::now().time_since_epoch()).count();

                SharedMemoryMessage message(MessageType::DATA, &data, sizeof(TestData));
                message.set_flag(MessageFlags::URGENT);

                if (producer_->send_message(message)) {
                    std::cout << "Sent structured data: id=" << data.id
                              << ", value=" << data.value << std::endl;
                }

                // 发送心跳
                if (counter % 10 == 0) {
                    producer_->send_heartbeat();
                    std::cout << "Sent heartbeat" << std::endl;
                }

                counter++;

                // 等待一段时间
                std::this_thread::sleep_for(std::chrono::milliseconds(500));

            } catch (const std::exception& e) {
                std::cerr << "Error in send loop: " << e.what() << std::endl;
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        }
    }

private:
    std::string name_;
    std::atomic<bool> running_;
    std::unique_ptr<SharedMemoryProducer> producer_;
    std::unique_ptr<std::thread> send_thread_;
};

class CrossLanguageTestConsumer {
public:
    CrossLanguageTestConsumer(const std::string& name = "CrossLangTest")
        : name_(name), running_(false) {
    }

    ~CrossLanguageTestConsumer() {
        stop();
    }

    bool start() {
        if (running_) {
            return true;
        }

        // 创建消费者
        consumer_ = create_consumer(name_, 1024 * 1024); // 1MB buffer
        if (!consumer_) {
            std::cerr << "Failed to create consumer" << std::endl;
            return false;
        }

        // 注册消息处理器
        consumer_->register_handler(MessageType::DATA,
            [this](const SharedMemoryMessage& msg) -> bool {
                std::cout << "Received data message: " << msg.get_payload_size() << " bytes" << std::endl;

                if (msg.get_payload_size() == sizeof(TestData)) {
                    const TestData* data = reinterpret_cast<const TestData*>(msg.get_payload());
                    std::cout << "  Parsed data: id=" << data->id
                              << ", value=" << data->value
                              << ", message=" << data->message
                              << ", timestamp=" << data->timestamp << std::endl;
                }

                return true;
            });

        consumer_->register_handler(MessageType::HEARTBEAT,
            [this](const SharedMemoryMessage& msg) -> bool {
                std::cout << "Received heartbeat from producer" << std::endl;
                return true;
            });

        std::cout << "✓ C++ Consumer connected to shared memory: " << name_ << std::endl;
        running_ = true;

        // 启动接收线程
        receive_thread_ = std::make_unique<std::thread>(&CrossLanguageTestConsumer::receive_loop, this);

        return true;
    }

    void stop() {
        if (!running_) {
            return;
        }

        running_ = false;

        if (receive_thread_ && receive_thread_->joinable()) {
            receive_thread_->join();
        }

        consumer_.reset();
        std::cout << "✓ C++ Consumer stopped" << std::endl;
    }

    bool is_running() const {
        return running_;
    }

private:
    void receive_loop() {
        while (running_) {
            try {
                // 尝试接收消息
                SharedMemoryMessage message;
                if (consumer_->receive_message(message, 100)) {
                    // 消息已经在处理器中处理
                    continue;
                }

                // 尝试接收字符串
                std::string text;
                if (consumer_->receive_string(text, 100)) {
                    std::cout << "Received string: " << text << std::endl;
                    continue;
                }

                // 尝试接收原始数据
                std::vector<uint8_t> data;
                if (consumer_->receive(data, 100)) {
                    std::cout << "Received raw data: " << data.size() << " bytes" << std::endl;
                    continue;
                }

                // 如果没有数据，短暂休眠
                std::this_thread::sleep_for(std::chrono::milliseconds(50));

            } catch (const std::exception& e) {
                std::cerr << "Error in receive loop: " << e.what() << std::endl;
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        }
    }

private:
    std::string name_;
    std::atomic<bool> running_;
    std::unique_ptr<SharedMemoryConsumer> consumer_;
    std::unique_ptr<std::thread> receive_thread_;
};

int main(int argc, char* argv[]) {
    std::cout << "BitRPC Cross-Language Shared Memory Test (C++)" << std::endl;
    std::cout << "=============================================" << std::endl;

    std::string mode = "producer"; // 默认为生产者
    std::string name = "CrossLangTest";

    // 解析命令行参数
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "--mode" && i + 1 < argc) {
            mode = argv[i + 1];
            i++;
        } else if (std::string(argv[i]) == "--name" && i + 1 < argc) {
            name = argv[i + 1];
            i++;
        } else if (std::string(argv[i]) == "--help" || std::string(argv[i]) == "-h") {
            std::cout << "Usage: " << argv[0] << " [options]" << std::endl;
            std::cout << "Options:" << std::endl;
            std::cout << "  --mode MODE    Set mode (producer/consumer)" << std::endl;
            std::cout << "  --name NAME    Set shared memory name" << std::endl;
            std::cout << "  --help, -h     Show this help" << std::endl;
            return 0;
        }
    }

    std::cout << "Mode: " << mode << std::endl;
    std::cout << "Shared Memory Name: " << name << std::endl;
    std::cout << std::endl;

    try {
        if (mode == "producer") {
            CrossLanguageTestProducer producer(name);
            if (!producer.start()) {
                std::cerr << "Failed to start producer" << std::endl;
                return 1;
            }

            std::cout << "Producer running. Press Ctrl+C to stop..." << std::endl;

            // 等待用户中断
            while (producer.is_running()) {
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }

        } else if (mode == "consumer") {
            CrossLanguageTestConsumer consumer(name);
            if (!consumer.start()) {
                std::cerr << "Failed to start consumer" << std::endl;
                return 1;
            }

            std::cout << "Consumer running. Press Ctrl+C to stop..." << std::endl;

            // 等待用户中断
            while (consumer.is_running()) {
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }

        } else {
            std::cerr << "Invalid mode: " << mode << std::endl;
            std::cerr << "Supported modes: producer, consumer" << std::endl;
            return 1;
        }

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    std::cout << "Test completed successfully!" << std::endl;
    return 0;
}